{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:33.647912Z",
     "iopub.status.busy": "2024-07-26T10:29:33.646909Z",
     "iopub.status.idle": "2024-07-26T10:29:34.985308Z",
     "shell.execute_reply": "2024-07-26T10:29:34.984312Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import random\n",
    "import pickle\n",
    "\n",
    "import edge2vec.transition3 as transitions\n",
    "import edge2vec.edge2vec3 as edge2vec\n",
    "\n",
    "import data_params as input_data_params"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify which dataset is used.\n",
    "*   1 (Original knowledge graph)\n",
    "*   2 (Restructured knowledge graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.004325Z",
     "iopub.status.busy": "2024-07-26T10:29:35.003324Z",
     "iopub.status.idle": "2024-07-26T10:29:35.017053Z",
     "shell.execute_reply": "2024-07-26T10:29:35.015019Z"
    }
   },
   "outputs": [],
   "source": [
    "dataset_prefix = input_data_params.dataset\n",
    "\n",
    "disease_prefix = input_data_params.disease\n",
    "assert disease_prefix == 'dmd' or 'hd' or 'oi'\n",
    "\n",
    "embedding_method = 'e2v'\n",
    "\n",
    "use_seed = input_data_params.e2v_seed\n",
    "\n",
    "if use_seed:\n",
    "    fixed_emb = '_seeded'\n",
    "else:\n",
    "    fixed_emb = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.021053Z",
     "iopub.status.busy": "2024-07-26T10:29:35.020053Z",
     "iopub.status.idle": "2024-07-26T10:29:35.033083Z",
     "shell.execute_reply": "2024-07-26T10:29:35.031065Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 already exists and will be used: C:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\hd\\g2_e2v\n"
     ]
    }
   ],
   "source": [
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output', disease_prefix)\n",
    "dataset_output_dir = os.path.join(curr_output_dir, f'{dataset_prefix}_{embedding_method}{fixed_emb}')\n",
    "\n",
    "if not os.path.exists(dataset_output_dir):\n",
    "    os.mkdir(dataset_output_dir)\n",
    "    print(f'Output folder for dataset {dataset_prefix} {disease_prefix} is created: {dataset_output_dir}')\n",
    "else:\n",
    "    print(f'Output folder for dataset {dataset_prefix} {disease_prefix} already exists and will be used: {dataset_output_dir}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.070676Z",
     "iopub.status.busy": "2024-07-26T10:29:35.070676Z",
     "iopub.status.idle": "2024-07-26T10:29:35.095548Z",
     "shell.execute_reply": "2024-07-26T10:29:35.093510Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for current run on dataset 2 is created: C:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\hd\\g2_e2v\\run_010\n"
     ]
    }
   ],
   "source": [
    "def create_run_folder(run_nr: int):\n",
    "    foldername = 'run_{:03d}'.format(run_nr)\n",
    "    run_path = os.path.join(dataset_output_dir, foldername)\n",
    "    os.mkdir(run_path)\n",
    "\n",
    "    print(f'Output folder for current run on dataset {dataset_prefix} {disease_prefix} is created: {run_path}')\n",
    "    return run_path\n",
    "\n",
    "run_folders_list = []\n",
    "\n",
    "for item in os.listdir(dataset_output_dir):\n",
    "    curr_path = os.path.join(dataset_output_dir, item)\n",
    "    if os.path.isdir(curr_path) and 'run' in item:\n",
    "        run_folders_list.append(item)\n",
    "\n",
    "curr_run_nr = 1\n",
    "if len(run_folders_list) > 0:\n",
    "    run_folders_list.sort(reverse=True)\n",
    "    latest_run = run_folders_list[0]\n",
    "    run_nr_str = latest_run.split('_')[1]\n",
    "\n",
    "    next_run_nr = int(run_nr_str) + 1\n",
    "    curr_run_nr = next_run_nr\n",
    "    output_path = create_run_folder(next_run_nr)\n",
    "else:\n",
    "    output_path = create_run_folder(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_seed_val(run_nr: int):\n",
    "    if use_seed:\n",
    "        seed_val = 10   # Fixed seed for each run\n",
    "        seed_txt_file = os.path.join(output_path, 'seed_value.txt')\n",
    "        with open(seed_txt_file, 'w') as f:\n",
    "            f.write(f'Seed value for this run is: {seed_val}')\n",
    "        print(f'File {seed_txt_file} has been saved storing the seed value used during the embedding step of current run.')\n",
    "\n",
    "        return seed_val\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "seed = set_seed_val(curr_run_nr)\n",
    "print(f'Current value of e2v embedding seed is {seed}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Nodes and Edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.099537Z",
     "iopub.status.busy": "2024-07-26T10:29:35.099537Z",
     "iopub.status.idle": "2024-07-26T10:29:35.141316Z",
     "shell.execute_reply": "2024-07-26T10:29:35.140281Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>dictyBase:DDB_G0282941</td>\n",
       "      <td>gene</td>\n",
       "      <td>DDB0185273</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>WBPhenotype:0001133</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>division axis defective</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>HP:0012005</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>Deja vu aura</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000043959</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000043959</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>HP:0001265</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>Hyporeflexia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14877</th>\n",
       "      <td>14877</td>\n",
       "      <td>MGI:2676637</td>\n",
       "      <td>gene</td>\n",
       "      <td>H2-M11</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878</th>\n",
       "      <td>14878</td>\n",
       "      <td>MP:0000378</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>absent hair follicles</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14879</th>\n",
       "      <td>14879</td>\n",
       "      <td>WormBase:WBGene00006446</td>\n",
       "      <td>gene</td>\n",
       "      <td>atx-3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14880</th>\n",
       "      <td>14880</td>\n",
       "      <td>MGI:2654144</td>\n",
       "      <td>gene</td>\n",
       "      <td>Ercc6l</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14881</th>\n",
       "      <td>14881</td>\n",
       "      <td>1912</td>\n",
       "      <td>drug</td>\n",
       "      <td>niclosamide</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14882 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_id                          id   semantic  \\\n",
       "0             0      dictyBase:DDB_G0282941       gene   \n",
       "1             1         WBPhenotype:0001133  phenotype   \n",
       "2             2                  HP:0012005  phenotype   \n",
       "3             3  ENSEMBL:ENSPTRG00000043959       gene   \n",
       "4             4                  HP:0001265  phenotype   \n",
       "...         ...                         ...        ...   \n",
       "14877     14877                 MGI:2676637       gene   \n",
       "14878     14878                  MP:0000378  phenotype   \n",
       "14879     14879     WormBase:WBGene00006446       gene   \n",
       "14880     14880                 MGI:2654144       gene   \n",
       "14881     14881                        1912       drug   \n",
       "\n",
       "                            label  semantic_id  \n",
       "0                      DDB0185273            4  \n",
       "1         division axis defective            8  \n",
       "2                    Deja vu aura            8  \n",
       "3      ENSEMBL:ENSPTRG00000043959            4  \n",
       "4                    Hyporeflexia            8  \n",
       "...                           ...          ...  \n",
       "14877                      H2-M11            4  \n",
       "14878       absent hair follicles            8  \n",
       "14879                       atx-3            4  \n",
       "14880                      Ercc6l            4  \n",
       "14881                 niclosamide            3  \n",
       "\n",
       "[14882 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/{disease_prefix}/{dataset_prefix}_{disease_prefix}_indexed_nodes.csv')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.145840Z",
     "iopub.status.busy": "2024-07-26T10:29:35.144832Z",
     "iopub.status.idle": "2024-07-26T10:29:35.425676Z",
     "shell.execute_reply": "2024-07-26T10:29:35.424690Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0000774</td>\n",
       "      <td>gametogenesis variant</td>\n",
       "      <td>8</td>\n",
       "      <td>9968</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0002188</td>\n",
       "      <td>male somatic gonad development variant</td>\n",
       "      <td>8</td>\n",
       "      <td>3292</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001224</td>\n",
       "      <td>axon outgrowth variant</td>\n",
       "      <td>8</td>\n",
       "      <td>8349</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:107812</td>\n",
       "      <td>Tubb5</td>\n",
       "      <td>4</td>\n",
       "      <td>4998</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001100</td>\n",
       "      <td>early embryonic lethal</td>\n",
       "      <td>8</td>\n",
       "      <td>6541</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230223</th>\n",
       "      <td>Q99460</td>\n",
       "      <td>26S proteasome non-ATPase regulatory subunit 1</td>\n",
       "      <td>5</td>\n",
       "      <td>4895</td>\n",
       "      <td>is product of</td>\n",
       "      <td>HGNC:9554</td>\n",
       "      <td>PSMD1</td>\n",
       "      <td>4</td>\n",
       "      <td>352</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230224</th>\n",
       "      <td>P53396</td>\n",
       "      <td>ATP-citrate synthase</td>\n",
       "      <td>5</td>\n",
       "      <td>1449</td>\n",
       "      <td>is product of</td>\n",
       "      <td>HGNC:115</td>\n",
       "      <td>ACLY</td>\n",
       "      <td>4</td>\n",
       "      <td>10260</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230225</th>\n",
       "      <td>Q60631</td>\n",
       "      <td>Growth factor receptor-bound protein 2</td>\n",
       "      <td>5</td>\n",
       "      <td>12956</td>\n",
       "      <td>is product of</td>\n",
       "      <td>MGI:95805</td>\n",
       "      <td>Grb2</td>\n",
       "      <td>4</td>\n",
       "      <td>6661</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230226</th>\n",
       "      <td>1980</td>\n",
       "      <td>octreotide</td>\n",
       "      <td>3</td>\n",
       "      <td>5596</td>\n",
       "      <td>targets</td>\n",
       "      <td>P35372</td>\n",
       "      <td>Mu-type opioid receptor</td>\n",
       "      <td>5</td>\n",
       "      <td>6566</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230227</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>5</td>\n",
       "      <td>6337</td>\n",
       "      <td>is product of</td>\n",
       "      <td>HGNC:3236</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>4</td>\n",
       "      <td>8284</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230228 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           head  \\\n",
       "0       WormBase:WBGene00006537   \n",
       "1       WormBase:WBGene00006537   \n",
       "2       WormBase:WBGene00006537   \n",
       "3       WormBase:WBGene00006537   \n",
       "4       WormBase:WBGene00006537   \n",
       "...                         ...   \n",
       "230223                   Q99460   \n",
       "230224                   P53396   \n",
       "230225                   Q60631   \n",
       "230226                     1980   \n",
       "230227                   P00533   \n",
       "\n",
       "                                            label_head  class_head  \\\n",
       "0                                                tbb-2           4   \n",
       "1                                                tbb-2           4   \n",
       "2                                                tbb-2           4   \n",
       "3                                                tbb-2           4   \n",
       "4                                                tbb-2           4   \n",
       "...                                                ...         ...   \n",
       "230223  26S proteasome non-ATPase regulatory subunit 1           5   \n",
       "230224                            ATP-citrate synthase           5   \n",
       "230225          Growth factor receptor-bound protein 2           5   \n",
       "230226                                      octreotide           3   \n",
       "230227                Epidermal growth factor receptor           5   \n",
       "\n",
       "        index_head                        relation                 tail  \\\n",
       "0            10272                causes condition  WBPhenotype:0000774   \n",
       "1            10272                causes condition  WBPhenotype:0002188   \n",
       "2            10272                causes condition  WBPhenotype:0001224   \n",
       "3            10272  in orthology relationship with           MGI:107812   \n",
       "4            10272                causes condition  WBPhenotype:0001100   \n",
       "...            ...                             ...                  ...   \n",
       "230223        4895                   is product of            HGNC:9554   \n",
       "230224        1449                   is product of             HGNC:115   \n",
       "230225       12956                   is product of            MGI:95805   \n",
       "230226        5596                         targets               P35372   \n",
       "230227        6337                   is product of            HGNC:3236   \n",
       "\n",
       "                                    label_tail  class_tail  index_tail  type  \n",
       "0                        gametogenesis variant           8        9968     0  \n",
       "1       male somatic gonad development variant           8        3292     0  \n",
       "2                       axon outgrowth variant           8        8349     0  \n",
       "3                                        Tubb5           4        4998     1  \n",
       "4                       early embryonic lethal           8        6541     0  \n",
       "...                                        ...         ...         ...   ...  \n",
       "230223                                   PSMD1           4         352    11  \n",
       "230224                                    ACLY           4       10260    11  \n",
       "230225                                    Grb2           4        6661    11  \n",
       "230226                 Mu-type opioid receptor           5        6566    12  \n",
       "230227                                    EGFR           4        8284    11  \n",
       "\n",
       "[230228 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/{disease_prefix}/{dataset_prefix}_{disease_prefix}_indexed_edges.csv')\n",
    "edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.429708Z",
     "iopub.status.busy": "2024-07-26T10:29:35.428684Z",
     "iopub.status.idle": "2024-07-26T10:29:35.441264Z",
     "shell.execute_reply": "2024-07-26T10:29:35.440263Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>1840</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000018930</td>\n",
       "      <td>gene</td>\n",
       "      <td>GIT1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_id                          id semantic label  semantic_id\n",
       "1840      1840  ENSEMBL:ENSCAFG00000018930     gene  GIT1            4"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.loc[nodes['index_id'] == 1840]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.443263Z",
     "iopub.status.busy": "2024-07-26T10:29:35.443263Z",
     "iopub.status.idle": "2024-07-26T10:29:35.471999Z",
     "shell.execute_reply": "2024-07-26T10:29:35.471010Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0000774</td>\n",
       "      <td>gametogenesis variant</td>\n",
       "      <td>8</td>\n",
       "      <td>9968</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0002188</td>\n",
       "      <td>male somatic gonad development variant</td>\n",
       "      <td>8</td>\n",
       "      <td>3292</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001224</td>\n",
       "      <td>axon outgrowth variant</td>\n",
       "      <td>8</td>\n",
       "      <td>8349</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:107812</td>\n",
       "      <td>Tubb5</td>\n",
       "      <td>4</td>\n",
       "      <td>4998</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001100</td>\n",
       "      <td>early embryonic lethal</td>\n",
       "      <td>8</td>\n",
       "      <td>6541</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000001379</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000001379</td>\n",
       "      <td>4</td>\n",
       "      <td>13971</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001102</td>\n",
       "      <td>mitotic spindle defective early emb</td>\n",
       "      <td>8</td>\n",
       "      <td>6832</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006533</td>\n",
       "      <td>tba-7</td>\n",
       "      <td>4</td>\n",
       "      <td>3902</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001037</td>\n",
       "      <td>sterile progeny</td>\n",
       "      <td>8</td>\n",
       "      <td>11086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WormBase:WBGene00006537</td>\n",
       "      <td>tbb-2</td>\n",
       "      <td>4</td>\n",
       "      <td>10272</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>ZP:0017152</td>\n",
       "      <td>germ cell development process quality, abnormal</td>\n",
       "      <td>8</td>\n",
       "      <td>5981</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      head label_head  class_head  index_head  \\\n",
       "0  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "1  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "2  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "3  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "4  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "5  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "6  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "7  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "8  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "9  WormBase:WBGene00006537      tbb-2           4       10272   \n",
       "\n",
       "                         relation                        tail  \\\n",
       "0                causes condition         WBPhenotype:0000774   \n",
       "1                causes condition         WBPhenotype:0002188   \n",
       "2                causes condition         WBPhenotype:0001224   \n",
       "3  in orthology relationship with                  MGI:107812   \n",
       "4                causes condition         WBPhenotype:0001100   \n",
       "5  in orthology relationship with  ENSEMBL:ENSSSCG00000001379   \n",
       "6                causes condition         WBPhenotype:0001102   \n",
       "7                  interacts with     WormBase:WBGene00006533   \n",
       "8                causes condition         WBPhenotype:0001037   \n",
       "9                causes condition                  ZP:0017152   \n",
       "\n",
       "                                        label_tail  class_tail  index_tail  \\\n",
       "0                            gametogenesis variant           8        9968   \n",
       "1           male somatic gonad development variant           8        3292   \n",
       "2                           axon outgrowth variant           8        8349   \n",
       "3                                            Tubb5           4        4998   \n",
       "4                           early embryonic lethal           8        6541   \n",
       "5                       ENSEMBL:ENSSSCG00000001379           4       13971   \n",
       "6              mitotic spindle defective early emb           8        6832   \n",
       "7                                            tba-7           4        3902   \n",
       "8                                  sterile progeny           8       11086   \n",
       "9  germ cell development process quality, abnormal           8        5981   \n",
       "\n",
       "   type  \n",
       "0     0  \n",
       "1     0  \n",
       "2     0  \n",
       "3     1  \n",
       "4     0  \n",
       "5     1  \n",
       "6     0  \n",
       "7     2  \n",
       "8     0  \n",
       "9     0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge2vec_df = edges.copy()\n",
    "edge2vec_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Edge2Vec Embedding Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize edge list with keys 'type' and 'weight'. The 'weight' of each edge is initialized to 1.0. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:35.476013Z",
     "iopub.status.busy": "2024-07-26T10:29:35.475004Z",
     "iopub.status.idle": "2024-07-26T10:29:37.492790Z",
     "shell.execute_reply": "2024-07-26T10:29:37.491768Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First node in graph: (10272, {})\n",
      "First edge in graph: (10272, 9968, {'type': 0, 'weight': 1.0})\n",
      "Total number of edges is 135988\n",
      "Total number of nodes is 14882\n"
     ]
    }
   ],
   "source": [
    "G = nx.from_pandas_edgelist(edge2vec_df, 'index_head', 'index_tail', 'type', create_using=nx.DiGraph(), edge_key= (('type', int),('id', int)))\n",
    "G = G.to_undirected()   # for the E2V implementation, use indirected graph\n",
    "for edge in G.edges():\n",
    "    G[edge[0]][edge[1]]['weight'] = 1.0\n",
    "    \n",
    "for node in G.nodes(data=True):\n",
    "    print('First node in graph:', node)\n",
    "    break\n",
    "\n",
    "for edge in G.edges(data=True):\n",
    "    print('First edge in graph:', edge)\n",
    "    break\n",
    "\n",
    "print('Total number of edges is {}'.format(G.number_of_edges()))\n",
    "print('Total number of nodes is {}'.format(G.number_of_nodes()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Count total number of edge types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:37.496803Z",
     "iopub.status.busy": "2024-07-26T10:29:37.496803Z",
     "iopub.status.idle": "2024-07-26T10:29:37.524339Z",
     "shell.execute_reply": "2024-07-26T10:29:37.523329Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 19 edge types\n"
     ]
    }
   ],
   "source": [
    "type_size = len(set(edge2vec_df['type']))\n",
    "print(f'There are {type_size} edge types')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtain Node Embeddings Using Edge2Vec"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define parameters and generate edge type transition matrix using Edge2Vec EM approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:37.528334Z",
     "iopub.status.busy": "2024-07-26T10:29:37.527334Z",
     "iopub.status.idle": "2024-07-26T10:29:37.539351Z",
     "shell.execute_reply": "2024-07-26T10:29:37.538339Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 5 128 0.5 1 10\n"
     ]
    }
   ],
   "source": [
    "e2v_params = {\n",
    "    'dmd': {\n",
    "        'prev': {\n",
    "            'epochs': 10,\n",
    "            'num_walks': 2,\n",
    "            'walk_length': 7,\n",
    "            'p': 0.70,\n",
    "            'q': 1.0,\n",
    "            'dim': 32\n",
    "        },\n",
    "        'restr': {\n",
    "            'epochs': 5,\n",
    "            'num_walks': 6,\n",
    "            'walk_length': 7,\n",
    "            'p': 0.75,\n",
    "            'q': 1.0,\n",
    "            'dim': 64\n",
    "        }\n",
    "    },\n",
    "    'hd': {\n",
    "        'prev': {\n",
    "            'epochs': 10,\n",
    "            'num_walks': 6,\n",
    "            'walk_length': 5,\n",
    "            'p': 0.75,\n",
    "            'q': 0.5,\n",
    "            'dim': 32\n",
    "        },\n",
    "        'restr': {\n",
    "            'epochs': 10,\n",
    "            'num_walks': 6,\n",
    "            'walk_length': 5,\n",
    "            'p': 0.5,\n",
    "            'q': 1.0,\n",
    "            'dim': 128\n",
    "        }\n",
    "    },\n",
    "    'oi': {\n",
    "        'prev': {\n",
    "            'epochs': 10,\n",
    "            'num_walks': 6,\n",
    "            'walk_length': 7,\n",
    "            'p': 1.0,\n",
    "            'q': 0.5,\n",
    "            'dim': 128\n",
    "        },\n",
    "        'restr': {\n",
    "            'epochs': 10,\n",
    "            'num_walks': 4,\n",
    "            'walk_length': 7,\n",
    "            'p': 0.5,\n",
    "            'q': 0.5,\n",
    "            'dim': 32\n",
    "        }\n",
    "    }\n",
    "}\n",
    "    \n",
    "directed = True\n",
    "e_step = 3\n",
    "\n",
    "params_to_use = e2v_params[disease_prefix][dataset_prefix]\n",
    "\n",
    "epochs = params_to_use['epochs']\n",
    "num_walks = params_to_use['num_walks']\n",
    "walk_length = params_to_use['walk_length']\n",
    "p = params_to_use['p']\n",
    "q = params_to_use['q']\n",
    "dim = params_to_use['dim']\n",
    "\n",
    "print(epochs, num_walks, walk_length, dim, p, q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:37.543342Z",
     "iopub.status.busy": "2024-07-26T10:29:37.542342Z",
     "iopub.status.idle": "2024-07-26T10:29:37.554868Z",
     "shell.execute_reply": "2024-07-26T10:29:37.553850Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_e2v_embeddings(save=False):\n",
    "    print('Initializing transition matrix...')\n",
    "    M = transitions.initialize_edge_type_matrix(type_size)\n",
    "\n",
    "    print('Updating transition matrix...')\n",
    "    for i in range(epochs):\n",
    "        walks = transitions.simulate_walks_1(G, num_walks, walk_length, M, directed, p, q, seed) # M step\n",
    "        print(f'{i}th iteration of updating matrix')\n",
    "        M = transitions.update_trans_matrix(walks, type_size, e_step) # E step\n",
    "\n",
    "    print(\"Finished generating values for transition matrix!\")\n",
    "    \n",
    "    print('Overview of resulting edge type transition matrix (float values have been truncated to fit the view!):')\n",
    "    print('\\n'.join(['\\t'.join([str(cell)[:5] for cell in row]) for row in M]))\n",
    "    \n",
    "    if save:\n",
    "        # Save transition matrix\n",
    "        np.save(f'{output_path}/transitionmatrix_{dataset_prefix}_{disease_prefix}.npy', M)\n",
    "    \n",
    "    print('Generate walks constrained by edge type transition matrix...')\n",
    "    walks = edge2vec.simulate_walks_2(G, num_walks, walk_length, M, p, q, seed)\n",
    "    \n",
    "    # Generate node embeddings using Word2Vec (skip-gram model) with as input the generated walks \n",
    "    window_size = walk_length - 1   # maximum distance between predicted and context node\n",
    "    workers = 8 # threads used\n",
    "\n",
    "    w2v_model = edge2vec.Word2Vec(walks, vector_size=dim, window=window_size, min_count=0, sg=1, workers=workers, epochs=epochs, seed=seed)\n",
    "    \n",
    "    word_vectors = w2v_model.wv\n",
    "    if save:\n",
    "        word_vectors.save(f'{output_path}/w2v_{dataset_prefix}_{disease_prefix}.dvectors')\n",
    "        \n",
    "    e2v_embedding = pd.DataFrame(columns = ['Node', 'Embedding'])\n",
    "    e2v_embedding_list = []\n",
    "    for _, key in enumerate(w2v_model.wv.index_to_key):\n",
    "        e2v_embedding.loc[int(key)] = pd.Series({'Node':int(key), 'Embedding':list(w2v_model.wv[key])})\n",
    "        e2v_embedding_list.append(list(w2v_model.wv[key]))\n",
    "        \n",
    "    e2v_embedding = e2v_embedding.sort_values('Node')\n",
    "    \n",
    "    return M, e2v_embedding, e2v_embedding_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-26T10:29:37.560850Z",
     "iopub.status.busy": "2024-07-26T10:29:37.559855Z",
     "iopub.status.idle": "2024-07-26T10:33:29.899021Z",
     "shell.execute_reply": "2024-07-26T10:33:29.897015Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing transition matrix...\n",
      "Updating transition matrix...\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "0th iteration of updating matrix\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\rosa-\\anaconda3\\envs\\xaifo\\lib\\site-packages\\scipy\\stats\\_morestats.py:3428: UserWarning: Sample size too small for normal approximation.\n",
      "  warnings.warn(\"Sample size too small for normal approximation.\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "1th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "2th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "3th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "4th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "5th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "6th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "7th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "8th iteration of updating matrix\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n",
      "9th iteration of updating matrix\n",
      "Finished generating values for transition matrix!\n",
      "Overview of resulting edge type transition matrix (float values have been truncated to fit the view!):\n",
      "0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t0.5\t1.0\t1.0\t1.0\t0.999\t1.0\t1.0\t0.999\t0.999\t0.5\n",
      "0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t0.5\t0.5\t1.0\t0.999\t1.0\t1.0\t1.0\t0.999\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t0.5\t0.5\t1.0\t0.999\t1.0\t1.0\t0.999\t0.5\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.999\t0.5\t0.999\t0.999\t1.0\t0.5\t0.999\t0.999\t0.999\t0.999\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t0.999\t0.5\t1.0\t1.0\t0.999\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.5\t1.0\t1.0\t1.0\t0.999\t1.0\t0.5\t1.0\t0.999\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.999\t0.5\t1.0\t0.999\t1.0\t0.999\t1.0\t1.0\t0.5\t0.999\t0.5\n",
      "1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t1.0\t0.999\t0.5\t0.999\t0.5\t1.0\t0.999\t0.999\t0.999\t0.999\t0.5\t0.5\n",
      "0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\t0.5\n",
      "Generate walks constrained by edge type transition matrix...\n",
      "Walk iteration:\n",
      "1 / 6\n",
      "2 / 6\n",
      "3 / 6\n",
      "4 / 6\n",
      "5 / 6\n",
      "6 / 6\n"
     ]
    }
   ],
   "source": [
    "get_e2v_embeddings(save=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
