{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook loads all edges and nodes of the relevant knowledge graph. New dataframes are created for the nodes and edges including index values for each node."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify which dataset is used.\n",
    "*   1 (Original knowledge graph)\n",
    "*   2 (Restructured knowledge graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 2\n",
    "assert dataset_nr == 1 or 2\n",
    "\n",
    "disease_prefix = 'hd'\n",
    "assert disease_prefix == 'dmd' or 'hd' or 'oi'\n",
    "\n",
    "if dataset_nr == 1:\n",
    "    prefix = 'prev'\n",
    "else:\n",
    "    prefix = 'restr'\n",
    "    \n",
    "FILE_EDGES = f'{prefix}_{disease_prefix}_kg_edges.csv'\n",
    "FILE_NODES = f'{prefix}_{disease_prefix}_kg_nodes.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load all edges of knowledge graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>relation_iri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c5f18132a9521c15c8abd0240da2c89c</td>\n",
       "      <td>HGNC:11280</td>\n",
       "      <td>HGNC:9539</td>\n",
       "      <td>RO:0002434</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2f4182da8cf6af6ca49912ea3ceabfee</td>\n",
       "      <td>FlyBase:FBgn0264855</td>\n",
       "      <td>MGI:101921</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5ca2f0ea7df2c1f6973774b80a1bcade</td>\n",
       "      <td>SGD:S000001709</td>\n",
       "      <td>WormBase:WBGene00001134</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0f2324357c6e3d10ae1e4573da055ae6</td>\n",
       "      <td>RGD:1595923</td>\n",
       "      <td>ENSEMBL:ENSDARG00000115781</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d7253620ceec8e39bcc3ab6cae816df6</td>\n",
       "      <td>WormBase:WBGene00012735</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>24902093f06314c017791733102c32b5</td>\n",
       "      <td>Xenbase:XB-GENE-486558</td>\n",
       "      <td>Xenbase:XB-GENE-967617</td>\n",
       "      <td>RO:0002434</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3fa27d22ca82a4a9a92d8691fda05019</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000017516</td>\n",
       "      <td>FlyBase:FBgn0034774</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>077bfbc256527fabd59ce71824c99298</td>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1cdd6ffe016bea4ca3ae5d547e89742f</td>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000029076</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>426d243f415d40b2833d6a3799625954</td>\n",
       "      <td>Xenbase:XB-GENE-952286</td>\n",
       "      <td>HGNC:1856</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id                     subject  \\\n",
       "0  c5f18132a9521c15c8abd0240da2c89c                  HGNC:11280   \n",
       "1  2f4182da8cf6af6ca49912ea3ceabfee         FlyBase:FBgn0264855   \n",
       "2  5ca2f0ea7df2c1f6973774b80a1bcade              SGD:S000001709   \n",
       "3  0f2324357c6e3d10ae1e4573da055ae6                 RGD:1595923   \n",
       "4  d7253620ceec8e39bcc3ab6cae816df6     WormBase:WBGene00012735   \n",
       "5  24902093f06314c017791733102c32b5      Xenbase:XB-GENE-486558   \n",
       "6  3fa27d22ca82a4a9a92d8691fda05019  ENSEMBL:ENSCAFG00000017516   \n",
       "7  077bfbc256527fabd59ce71824c99298  ENSEMBL:ENSECAG00000022119   \n",
       "8  1cdd6ffe016bea4ca3ae5d547e89742f  ENSEMBL:ENSRNOG00000049403   \n",
       "9  426d243f415d40b2833d6a3799625954      Xenbase:XB-GENE-952286   \n",
       "\n",
       "                       object    relation_id                  relation_label  \\\n",
       "0                   HGNC:9539     RO:0002434                  interacts with   \n",
       "1                  MGI:101921  RO:HOM0000017  in orthology relationship with   \n",
       "2     WormBase:WBGene00001134  RO:HOM0000017  in orthology relationship with   \n",
       "3  ENSEMBL:ENSDARG00000115781  RO:HOM0000017  in orthology relationship with   \n",
       "4              FlyBase:CG5669  RO:HOM0000017  in orthology relationship with   \n",
       "5      Xenbase:XB-GENE-967617     RO:0002434                  interacts with   \n",
       "6         FlyBase:FBgn0034774  RO:HOM0000017  in orthology relationship with   \n",
       "7  ENSEMBL:ENSMODG00000015091  RO:HOM0000017  in orthology relationship with   \n",
       "8  ENSEMBL:ENSCAFG00000029076  RO:HOM0000017  in orthology relationship with   \n",
       "9                   HGNC:1856  RO:HOM0000017  in orthology relationship with   \n",
       "\n",
       "                                   relation_iri  \n",
       "0                                           NaN  \n",
       "1                                           NaN  \n",
       "2                                           NaN  \n",
       "3                                           NaN  \n",
       "4                                           NaN  \n",
       "5                                           NaN  \n",
       "6                                           NaN  \n",
       "7  http://purl.obolibrary.org/obo/RO_HOM0000017  \n",
       "8                                           NaN  \n",
       "9  http://purl.obolibrary.org/obo/RO_HOM0000017  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'../output/{disease_prefix}/{FILE_EDGES}', header=0)\n",
    "edges.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract the relevant columns and rename them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:11280</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:9539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FlyBase:FBgn0264855</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:101921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGD:S000001709</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00001134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGD:1595923</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSDARG00000115781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00012735</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Xenbase:XB-GENE-486558</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>Xenbase:XB-GENE-967617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000017516</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0034774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000029076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Xenbase:XB-GENE-952286</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:1856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         head                        relation  \\\n",
       "0                  HGNC:11280                  interacts with   \n",
       "1         FlyBase:FBgn0264855  in orthology relationship with   \n",
       "2              SGD:S000001709  in orthology relationship with   \n",
       "3                 RGD:1595923  in orthology relationship with   \n",
       "4     WormBase:WBGene00012735  in orthology relationship with   \n",
       "5      Xenbase:XB-GENE-486558                  interacts with   \n",
       "6  ENSEMBL:ENSCAFG00000017516  in orthology relationship with   \n",
       "7  ENSEMBL:ENSECAG00000022119  in orthology relationship with   \n",
       "8  ENSEMBL:ENSRNOG00000049403  in orthology relationship with   \n",
       "9      Xenbase:XB-GENE-952286  in orthology relationship with   \n",
       "\n",
       "                         tail  \n",
       "0                   HGNC:9539  \n",
       "1                  MGI:101921  \n",
       "2     WormBase:WBGene00001134  \n",
       "3  ENSEMBL:ENSDARG00000115781  \n",
       "4              FlyBase:CG5669  \n",
       "5      Xenbase:XB-GENE-967617  \n",
       "6         FlyBase:FBgn0034774  \n",
       "7  ENSEMBL:ENSMODG00000015091  \n",
       "8  ENSEMBL:ENSCAFG00000029076  \n",
       "9                   HGNC:1856  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = edges[['subject', 'relation_label', 'object']]\n",
    "edges.rename(columns={'subject': 'head', 'relation_label': 'relation', 'object': 'tail'}, inplace=True)\n",
    "edges.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_int_str(value):\n",
    "    try:\n",
    "        return str(int(float(value)))\n",
    "    except ValueError:\n",
    "        return value\n",
    "\n",
    "edges['head'] = edges['head'].apply(convert_to_int_str)\n",
    "edges['tail'] = edges['tail'].apply(convert_to_int_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 230228 edges\n"
     ]
    }
   ],
   "source": [
    "print(f'There are {edges.shape[0]} edges')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load all nodes of knowledge graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>iri</th>\n",
       "      <th>semantic</th>\n",
       "      <th>index_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSEMBL:ENSSSCG00000033987</td>\n",
       "      <td>UBE2E3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MP:0000790</td>\n",
       "      <td>abnormal stratification in cerebral cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZP:0011626</td>\n",
       "      <td>cell trunk apoptotic, abnormal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZP:0019272</td>\n",
       "      <td>pronephric nephron tubule epithelial cell diff...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MP:0009154</td>\n",
       "      <td>pancreatic acinar hypoplasia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WormBase:WBGene00002583</td>\n",
       "      <td>let-363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZP:0000943</td>\n",
       "      <td>eye morphology, abnormal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSEMBL:ENSSSCG00000012913</td>\n",
       "      <td>PPP1CA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id  \\\n",
       "0  ENSEMBL:ENSSSCG00000033987   \n",
       "1                  MP:0000790   \n",
       "2                  ZP:0011626   \n",
       "3                  ZP:0019272   \n",
       "4  ENSEMBL:ENSACAG00000007327   \n",
       "5                  MP:0009154   \n",
       "6  ENSEMBL:ENSOANG00000047629   \n",
       "7     WormBase:WBGene00002583   \n",
       "8                  ZP:0000943   \n",
       "9  ENSEMBL:ENSSSCG00000012913   \n",
       "\n",
       "                                               label  iri   semantic  index_id  \n",
       "0                                             UBE2E3  NaN       gene         0  \n",
       "1         abnormal stratification in cerebral cortex  NaN  phenotype         1  \n",
       "2                     cell trunk apoptotic, abnormal  NaN  phenotype         2  \n",
       "3  pronephric nephron tubule epithelial cell diff...  NaN  phenotype         3  \n",
       "4                         ENSEMBL:ENSACAG00000007327  NaN       gene         4  \n",
       "5                       pancreatic acinar hypoplasia  NaN  phenotype         5  \n",
       "6                         ENSEMBL:ENSOANG00000047629  NaN       gene         6  \n",
       "7                                            let-363  NaN       gene         7  \n",
       "8                           eye morphology, abnormal  NaN  phenotype         8  \n",
       "9                                             PPP1CA  NaN       gene         9  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'../output/{disease_prefix}/{FILE_NODES}', header=0)\n",
    "nodes['index_id'] = nodes.index\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract relevant columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000033987</td>\n",
       "      <td>gene</td>\n",
       "      <td>UBE2E3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MP:0000790</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal stratification in cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ZP:0011626</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cell trunk apoptotic, abnormal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ZP:0019272</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pronephric nephron tubule epithelial cell diff...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>MP:0009154</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pancreatic acinar hypoplasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>WormBase:WBGene00002583</td>\n",
       "      <td>gene</td>\n",
       "      <td>let-363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>ZP:0000943</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>eye morphology, abnormal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000012913</td>\n",
       "      <td>gene</td>\n",
       "      <td>PPP1CA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_id                          id   semantic  \\\n",
       "0         0  ENSEMBL:ENSSSCG00000033987       gene   \n",
       "1         1                  MP:0000790  phenotype   \n",
       "2         2                  ZP:0011626  phenotype   \n",
       "3         3                  ZP:0019272  phenotype   \n",
       "4         4  ENSEMBL:ENSACAG00000007327       gene   \n",
       "5         5                  MP:0009154  phenotype   \n",
       "6         6  ENSEMBL:ENSOANG00000047629       gene   \n",
       "7         7     WormBase:WBGene00002583       gene   \n",
       "8         8                  ZP:0000943  phenotype   \n",
       "9         9  ENSEMBL:ENSSSCG00000012913       gene   \n",
       "\n",
       "                                               label  \n",
       "0                                             UBE2E3  \n",
       "1         abnormal stratification in cerebral cortex  \n",
       "2                     cell trunk apoptotic, abnormal  \n",
       "3  pronephric nephron tubule epithelial cell diff...  \n",
       "4                         ENSEMBL:ENSACAG00000007327  \n",
       "5                       pancreatic acinar hypoplasia  \n",
       "6                         ENSEMBL:ENSOANG00000047629  \n",
       "7                                            let-363  \n",
       "8                           eye morphology, abnormal  \n",
       "9                                             PPP1CA  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = nodes[['index_id', 'id', 'semantic', 'label']]\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 14882 nodes\n"
     ]
    }
   ],
   "source": [
    "print(f'There are {nodes.shape[0]} nodes')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert semantic groups from categorical to numerical values (`semantic_id`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000033987</td>\n",
       "      <td>gene</td>\n",
       "      <td>UBE2E3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MP:0000790</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal stratification in cerebral cortex</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ZP:0011626</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cell trunk apoptotic, abnormal</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ZP:0019272</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pronephric nephron tubule epithelial cell diff...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000007327</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>MP:0009154</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pancreatic acinar hypoplasia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSOANG00000047629</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>WormBase:WBGene00002583</td>\n",
       "      <td>gene</td>\n",
       "      <td>let-363</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>ZP:0000943</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>eye morphology, abnormal</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000012913</td>\n",
       "      <td>gene</td>\n",
       "      <td>PPP1CA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_id                          id   semantic  \\\n",
       "0         0  ENSEMBL:ENSSSCG00000033987       gene   \n",
       "1         1                  MP:0000790  phenotype   \n",
       "2         2                  ZP:0011626  phenotype   \n",
       "3         3                  ZP:0019272  phenotype   \n",
       "4         4  ENSEMBL:ENSACAG00000007327       gene   \n",
       "5         5                  MP:0009154  phenotype   \n",
       "6         6  ENSEMBL:ENSOANG00000047629       gene   \n",
       "7         7     WormBase:WBGene00002583       gene   \n",
       "8         8                  ZP:0000943  phenotype   \n",
       "9         9  ENSEMBL:ENSSSCG00000012913       gene   \n",
       "\n",
       "                                               label  semantic_id  \n",
       "0                                             UBE2E3            4  \n",
       "1         abnormal stratification in cerebral cortex            8  \n",
       "2                     cell trunk apoptotic, abnormal            8  \n",
       "3  pronephric nephron tubule epithelial cell diff...            8  \n",
       "4                         ENSEMBL:ENSACAG00000007327            4  \n",
       "5                       pancreatic acinar hypoplasia            8  \n",
       "6                         ENSEMBL:ENSOANG00000047629            4  \n",
       "7                                            let-363            4  \n",
       "8                           eye morphology, abnormal            8  \n",
       "9                                             PPP1CA            4  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes['semantic'] = nodes['semantic'].astype('category')\n",
    "nodes['semantic_id'] = nodes['semantic'].cat.codes\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a dictionary that can be used as mapping between `semantic_id` and `semantic`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'biological artifact',\n",
       " 1: 'biological process',\n",
       " 2: 'disease',\n",
       " 3: 'drug',\n",
       " 4: 'gene',\n",
       " 5: 'gene product',\n",
       " 6: 'genotype',\n",
       " 7: 'molecular function',\n",
       " 8: 'phenotype',\n",
       " 9: 'taxon',\n",
       " 10: 'variant'}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_semantic_dict = dict(enumerate(nodes['semantic'].cat.categories))\n",
    "node_semantic_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save new dataframe containing all nodes into csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes.to_csv(f'../output/{disease_prefix}/{prefix}_{disease_prefix}_indexed_nodes.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Joining Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge dataframes with nodes and edges in order to join semantic classes and indices of nodes with heads and tails of the edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:11280</td>\n",
       "      <td>SQSTM1</td>\n",
       "      <td>4</td>\n",
       "      <td>9467</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:9539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FlyBase:FBgn0264855</td>\n",
       "      <td>AP-2alpha</td>\n",
       "      <td>4</td>\n",
       "      <td>6589</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:101921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGD:S000001709</td>\n",
       "      <td>VPS1</td>\n",
       "      <td>4</td>\n",
       "      <td>9926</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00001134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGD:1595923</td>\n",
       "      <td>RT1-A2</td>\n",
       "      <td>4</td>\n",
       "      <td>9824</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSDARG00000115781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00012735</td>\n",
       "      <td>sptf-3</td>\n",
       "      <td>4</td>\n",
       "      <td>1165</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Xenbase:XB-GENE-486558</td>\n",
       "      <td>pias1</td>\n",
       "      <td>4</td>\n",
       "      <td>8551</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>Xenbase:XB-GENE-967617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000017516</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>4</td>\n",
       "      <td>10763</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0034774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>4</td>\n",
       "      <td>8715</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>4</td>\n",
       "      <td>2590</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000029076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Xenbase:XB-GENE-952286</td>\n",
       "      <td>cenpe</td>\n",
       "      <td>4</td>\n",
       "      <td>8623</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:1856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         head                  label_head  class_head  \\\n",
       "0                  HGNC:11280                      SQSTM1           4   \n",
       "1         FlyBase:FBgn0264855                   AP-2alpha           4   \n",
       "2              SGD:S000001709                        VPS1           4   \n",
       "3                 RGD:1595923                      RT1-A2           4   \n",
       "4     WormBase:WBGene00012735                      sptf-3           4   \n",
       "5      Xenbase:XB-GENE-486558                       pias1           4   \n",
       "6  ENSEMBL:ENSCAFG00000017516                       CALM1           4   \n",
       "7  ENSEMBL:ENSECAG00000022119  ENSEMBL:ENSECAG00000022119           4   \n",
       "8  ENSEMBL:ENSRNOG00000049403  ENSEMBL:ENSRNOG00000049403           4   \n",
       "9      Xenbase:XB-GENE-952286                       cenpe           4   \n",
       "\n",
       "   index_head                        relation                        tail  \n",
       "0        9467                  interacts with                   HGNC:9539  \n",
       "1        6589  in orthology relationship with                  MGI:101921  \n",
       "2        9926  in orthology relationship with     WormBase:WBGene00001134  \n",
       "3        9824  in orthology relationship with  ENSEMBL:ENSDARG00000115781  \n",
       "4        1165  in orthology relationship with              FlyBase:CG5669  \n",
       "5        8551                  interacts with      Xenbase:XB-GENE-967617  \n",
       "6       10763  in orthology relationship with         FlyBase:FBgn0034774  \n",
       "7        8715  in orthology relationship with  ENSEMBL:ENSMODG00000015091  \n",
       "8        2590  in orthology relationship with  ENSEMBL:ENSCAFG00000029076  \n",
       "9        8623  in orthology relationship with                   HGNC:1856  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df = pd.merge(edges, nodes,\n",
    "                 left_on='head', right_on='id',\n",
    "                 how='inner')\n",
    "kg_df = kg_df[['head', 'label', 'semantic_id', 'index_id', 'relation', 'tail']]\n",
    "kg_df.rename(columns={'label': 'label_head', 'semantic_id': 'class_head', 'index_id': 'index_head'}, inplace=True)\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:11280</td>\n",
       "      <td>SQSTM1</td>\n",
       "      <td>4</td>\n",
       "      <td>9467</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:9539</td>\n",
       "      <td>PSMB2</td>\n",
       "      <td>4</td>\n",
       "      <td>12355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FlyBase:FBgn0264855</td>\n",
       "      <td>AP-2alpha</td>\n",
       "      <td>4</td>\n",
       "      <td>6589</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:101921</td>\n",
       "      <td>Ap2a1</td>\n",
       "      <td>4</td>\n",
       "      <td>10645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGD:S000001709</td>\n",
       "      <td>VPS1</td>\n",
       "      <td>4</td>\n",
       "      <td>9926</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00001134</td>\n",
       "      <td>eat-3</td>\n",
       "      <td>4</td>\n",
       "      <td>8881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGD:1595923</td>\n",
       "      <td>RT1-A2</td>\n",
       "      <td>4</td>\n",
       "      <td>9824</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSDARG00000115781</td>\n",
       "      <td>CR339041.3</td>\n",
       "      <td>4</td>\n",
       "      <td>14166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00012735</td>\n",
       "      <td>sptf-3</td>\n",
       "      <td>4</td>\n",
       "      <td>1165</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "      <td>4</td>\n",
       "      <td>3446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Xenbase:XB-GENE-486558</td>\n",
       "      <td>pias1</td>\n",
       "      <td>4</td>\n",
       "      <td>8551</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>Xenbase:XB-GENE-967617</td>\n",
       "      <td>cetn1</td>\n",
       "      <td>4</td>\n",
       "      <td>6209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000017516</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>4</td>\n",
       "      <td>10763</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0034774</td>\n",
       "      <td>CG13526</td>\n",
       "      <td>4</td>\n",
       "      <td>1670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>4</td>\n",
       "      <td>8715</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "      <td>4</td>\n",
       "      <td>3536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>4</td>\n",
       "      <td>2590</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000029076</td>\n",
       "      <td>CCDC126</td>\n",
       "      <td>4</td>\n",
       "      <td>10793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Xenbase:XB-GENE-952286</td>\n",
       "      <td>cenpe</td>\n",
       "      <td>4</td>\n",
       "      <td>8623</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:1856</td>\n",
       "      <td>CENPE</td>\n",
       "      <td>4</td>\n",
       "      <td>3187</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         head                  label_head  class_head  \\\n",
       "0                  HGNC:11280                      SQSTM1           4   \n",
       "1         FlyBase:FBgn0264855                   AP-2alpha           4   \n",
       "2              SGD:S000001709                        VPS1           4   \n",
       "3                 RGD:1595923                      RT1-A2           4   \n",
       "4     WormBase:WBGene00012735                      sptf-3           4   \n",
       "5      Xenbase:XB-GENE-486558                       pias1           4   \n",
       "6  ENSEMBL:ENSCAFG00000017516                       CALM1           4   \n",
       "7  ENSEMBL:ENSECAG00000022119  ENSEMBL:ENSECAG00000022119           4   \n",
       "8  ENSEMBL:ENSRNOG00000049403  ENSEMBL:ENSRNOG00000049403           4   \n",
       "9      Xenbase:XB-GENE-952286                       cenpe           4   \n",
       "\n",
       "   index_head                        relation                        tail  \\\n",
       "0        9467                  interacts with                   HGNC:9539   \n",
       "1        6589  in orthology relationship with                  MGI:101921   \n",
       "2        9926  in orthology relationship with     WormBase:WBGene00001134   \n",
       "3        9824  in orthology relationship with  ENSEMBL:ENSDARG00000115781   \n",
       "4        1165  in orthology relationship with              FlyBase:CG5669   \n",
       "5        8551                  interacts with      Xenbase:XB-GENE-967617   \n",
       "6       10763  in orthology relationship with         FlyBase:FBgn0034774   \n",
       "7        8715  in orthology relationship with  ENSEMBL:ENSMODG00000015091   \n",
       "8        2590  in orthology relationship with  ENSEMBL:ENSCAFG00000029076   \n",
       "9        8623  in orthology relationship with                   HGNC:1856   \n",
       "\n",
       "                   label_tail  class_tail  index_tail  \n",
       "0                       PSMB2           4       12355  \n",
       "1                       Ap2a1           4       10645  \n",
       "2                       eat-3           4        8881  \n",
       "3                  CR339041.3           4       14166  \n",
       "4              FlyBase:CG5669           4        3446  \n",
       "5                       cetn1           4        6209  \n",
       "6                     CG13526           4        1670  \n",
       "7  ENSEMBL:ENSMODG00000015091           4        3536  \n",
       "8                     CCDC126           4       10793  \n",
       "9                       CENPE           4        3187  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df = pd.merge(kg_df, nodes,\n",
    "                 left_on='tail', right_on='id',\n",
    "                 how='left')\n",
    "kg_df = kg_df[['head', 'label_head', 'class_head', 'index_head', 'relation', 'tail', 'label', 'semantic_id', 'index_id']]\n",
    "kg_df.rename(columns={'label': 'label_tail', 'semantic_id': 'class_tail', 'index_id': 'index_tail'}, inplace=True)\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Encode string values of relation labels as numeric values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:11280</td>\n",
       "      <td>SQSTM1</td>\n",
       "      <td>4</td>\n",
       "      <td>9467</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:9539</td>\n",
       "      <td>PSMB2</td>\n",
       "      <td>4</td>\n",
       "      <td>12355</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FlyBase:FBgn0264855</td>\n",
       "      <td>AP-2alpha</td>\n",
       "      <td>4</td>\n",
       "      <td>6589</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:101921</td>\n",
       "      <td>Ap2a1</td>\n",
       "      <td>4</td>\n",
       "      <td>10645</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGD:S000001709</td>\n",
       "      <td>VPS1</td>\n",
       "      <td>4</td>\n",
       "      <td>9926</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00001134</td>\n",
       "      <td>eat-3</td>\n",
       "      <td>4</td>\n",
       "      <td>8881</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGD:1595923</td>\n",
       "      <td>RT1-A2</td>\n",
       "      <td>4</td>\n",
       "      <td>9824</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSDARG00000115781</td>\n",
       "      <td>CR339041.3</td>\n",
       "      <td>4</td>\n",
       "      <td>14166</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00012735</td>\n",
       "      <td>sptf-3</td>\n",
       "      <td>4</td>\n",
       "      <td>1165</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "      <td>FlyBase:CG5669</td>\n",
       "      <td>4</td>\n",
       "      <td>3446</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Xenbase:XB-GENE-486558</td>\n",
       "      <td>pias1</td>\n",
       "      <td>4</td>\n",
       "      <td>8551</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>Xenbase:XB-GENE-967617</td>\n",
       "      <td>cetn1</td>\n",
       "      <td>4</td>\n",
       "      <td>6209</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000017516</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>4</td>\n",
       "      <td>10763</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0034774</td>\n",
       "      <td>CG13526</td>\n",
       "      <td>4</td>\n",
       "      <td>1670</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>ENSEMBL:ENSECAG00000022119</td>\n",
       "      <td>4</td>\n",
       "      <td>8715</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "      <td>ENSEMBL:ENSMODG00000015091</td>\n",
       "      <td>4</td>\n",
       "      <td>3536</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>ENSEMBL:ENSRNOG00000049403</td>\n",
       "      <td>4</td>\n",
       "      <td>2590</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000029076</td>\n",
       "      <td>CCDC126</td>\n",
       "      <td>4</td>\n",
       "      <td>10793</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Xenbase:XB-GENE-952286</td>\n",
       "      <td>cenpe</td>\n",
       "      <td>4</td>\n",
       "      <td>8623</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:1856</td>\n",
       "      <td>CENPE</td>\n",
       "      <td>4</td>\n",
       "      <td>3187</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         head                  label_head  class_head  \\\n",
       "0                  HGNC:11280                      SQSTM1           4   \n",
       "1         FlyBase:FBgn0264855                   AP-2alpha           4   \n",
       "2              SGD:S000001709                        VPS1           4   \n",
       "3                 RGD:1595923                      RT1-A2           4   \n",
       "4     WormBase:WBGene00012735                      sptf-3           4   \n",
       "5      Xenbase:XB-GENE-486558                       pias1           4   \n",
       "6  ENSEMBL:ENSCAFG00000017516                       CALM1           4   \n",
       "7  ENSEMBL:ENSECAG00000022119  ENSEMBL:ENSECAG00000022119           4   \n",
       "8  ENSEMBL:ENSRNOG00000049403  ENSEMBL:ENSRNOG00000049403           4   \n",
       "9      Xenbase:XB-GENE-952286                       cenpe           4   \n",
       "\n",
       "   index_head                        relation                        tail  \\\n",
       "0        9467                  interacts with                   HGNC:9539   \n",
       "1        6589  in orthology relationship with                  MGI:101921   \n",
       "2        9926  in orthology relationship with     WormBase:WBGene00001134   \n",
       "3        9824  in orthology relationship with  ENSEMBL:ENSDARG00000115781   \n",
       "4        1165  in orthology relationship with              FlyBase:CG5669   \n",
       "5        8551                  interacts with      Xenbase:XB-GENE-967617   \n",
       "6       10763  in orthology relationship with         FlyBase:FBgn0034774   \n",
       "7        8715  in orthology relationship with  ENSEMBL:ENSMODG00000015091   \n",
       "8        2590  in orthology relationship with  ENSEMBL:ENSCAFG00000029076   \n",
       "9        8623  in orthology relationship with                   HGNC:1856   \n",
       "\n",
       "                   label_tail  class_tail  index_tail  type  \n",
       "0                       PSMB2           4       12355     0  \n",
       "1                       Ap2a1           4       10645     1  \n",
       "2                       eat-3           4        8881     1  \n",
       "3                  CR339041.3           4       14166     1  \n",
       "4              FlyBase:CG5669           4        3446     1  \n",
       "5                       cetn1           4        6209     0  \n",
       "6                     CG13526           4        1670     1  \n",
       "7  ENSEMBL:ENSMODG00000015091           4        3536     1  \n",
       "8                     CCDC126           4       10793     1  \n",
       "9                       CENPE           4        3187     1  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df['type'], relation_labels = pd.factorize(kg_df['relation'])\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['interacts with', 'in orthology relationship with', 'causes condition',\n",
       "       'is variant in', 'enables', 'found in', 'contributes to condition',\n",
       "       'colocalizes with', 'has role in modeling', 'expresses gene',\n",
       "       'associated with phenotype', 'is of', 'targets',\n",
       "       'is substance that treats', 'has affected feature', 'is product of',\n",
       "       'is allele of', 'has genotype', 'involved in'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relation_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save new dataframe containing all edges into csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "kg_df.to_csv(f'../output/{disease_prefix}/{prefix}_{disease_prefix}_indexed_edges.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
